### DISCUSSION TITLE
Share your Ideas

### DISCUSSION THREAD

**Discussion by ZarehD** [ZarehD/AspNetStatic#5] [Ideas]
Please use this discussion thread to share your ideas for AspNetStatic.

Thank you.

**Comment by westonsoftware** [comment]
Hi, I was thinking about calling GenerateStaticPages() from an IHostedService background class as part of bigger project, does that seem possible?
(thanks for sharing this project)
--Andy

**Reply by ZarehD** [reply]
As long as you have a reference to a valid IHost instance, I don't see any significance to where this is called from. The only thing to be aware of is that this needs to be called before `app.Run()` so that it can get the app-started event when the host app starts, otherwise it won't start the page-generation process (unless, ofc, you can somehow manually trigger that event yourself).

**Comment by timalmond** [comment]
Hi. I tried this and it seemed to work nicely.

Two things I was wondering about:-

1. Could the code also copy linked files into the output, for example images?
2. Is there a way to remove things if it's been statically rendered? I'm thinking things like login links which obviously aren't going to work on the static site?

Thanks again. I've taken Miniblog, added pages and the idea is that I can use it to create S3 blogs with your library.

**Reply by ZarehD** [reply]
Hi, @timalmond.

I'm glad you find AspNetStatic useful. Don't forget to give the repo a star. It would be much appreciated.

On item 1, copying related assets to the output location, I recommend defining a post-build action for the project, or use a Gulp script bound to a build event (see Task Runner).

As for item 2, dynamically altering page content when generating for static output, you can use the usual Razor flow control syntax in your page to include/exclude content based on a flag.
```Razor
@using AspNetStatic
@{
    var isGeneratingForStatic = this.Request.IsAspNetStaticRequest();
}
@if (!isGeneratingForStatic)
{
    <div>
        Content I want to output only when not generating a static page....
    </div>
}
```

**Comment by timalmond** [comment]
I thought of something else: is there a way to trigger a rebuild from within the code? I'd like to get my code to regenerate on an update. Do I inject the service or something?

Thanks. Have starred the repo and will be talking about this for a talk

**Reply by ZarehD** [reply]
There's currently no direct way to trigger re-generation. I will look at providing something for this in the near future. In the mean time, you can specify a non-null value for the `regenerationInterval` parameter in the `GenerateStaticPages()` call to configure periodic re-generation.

**Reply by ZarehD** [reply]
Hi, @timalmond.

It turns out '_in the near future_' is actually today!

I added a new extension method, `GenerateStaticPagesNow()`, that initiates immediate static page generation. Note that you'll need to inject the `IHost` service (`WebApplication` instance) into the code where you want to call this from, and that the app is fully loaded/started and ready to serve pages when the method is called.

Install the new `0.17.0` release to get the code.

**Reply by westonsoftware** [reply]
@ZarehD hey thanks!  Where can I get the IHost?  Not sure how to get it from services, do I add it? or setup a global?

**Reply by ZarehD** [reply]
Hi, @westonsoftware.

The web-app instance is already registered in the DI container, so just add an `IHost` parameter to the ctor of the class where you want to use this.

**Reply by PaifferDev** [reply]
@ZarehD Hello, thank you very much, this tool is very useful, this new GenerateStaticPagesNow() option is very useful for me since I want that when a user who is making changes to the web, when he believes that it is ready to publish, this function is called and at that moment it generates the static files...
I have 2 doubts that I don't know if you can clarify:
1- it is possible not to generate static files when compile, only do it when GenerateStaticPagesNow() is executed
2- Right now, using HtmlAgilityPack, I browse the img-type nodes and if they don't have the loading="lazy" property, add it, can a functionality like this be extended in the Optimization process?
.Greetings

**Reply by ZarehD** [reply]
Hello @FPaiffer,

Thank you for the feedback.

Yes, you can omit the `GenerateStaticPages` call in the app startup code. The purpose of that function is to allow AspNetStatic to perform static generation (one-time or periodic) once the app is ready to serve pages. You don't have to initiate static generation that way; you can instead (or in addition) use the `GenerateStaticPagesNow` call from somewhere in your code (once the app is up and running).

With regard to the second question, if I understand it correctly, you want AspNetStatic to add an attribute (if missing) to an image tag as part of content optimization. Unfortunately, that's something that's out-of-scope for AspNetStatic. However, since AspNetStatic uses `WebMarkupMin` to implement content optimization, and given that you can specify/use custom optimizers with AspNetStatic (see bottom of Readme file to learn how), you should just extend one of the WebMarkupMin optimizers to implement what you want.
